test/gtest/ucp/test_ucp_wakeup.cc

```diff
@@ -174,3 +175,266 @@ UCS_TEST_P(test_ucp_wakeup, signal)
 
 UCP_INSTANTIATE_TEST_CASE(test_ucp_wakeup)
 
+class test_ucp_wakeup_external_epollfd : public test_ucp_wakeup {
+public:
+    virtual ucp_worker_params_t get_worker_params() {
+        ucp_worker_params_t params = test_ucp_wakeup::get_worker_params();
+        params.field_mask |= UCP_WORKER_PARAM_FIELD_EPOLL;
+        params.epoll.epoll_fd        = m_epfd;
+        params.epoll.epoll_data.u32 = EP_DATA32;
+        return params;
+    }
+
+protected:
+    enum {
+        EP_DATA32 = 0x1337
+    };
+
+    virtual void init() {
+        m_epfd = epoll_create(1);
+        ASSERT_GE(m_epfd, 0);
+        test_ucp_wakeup::init();
+    }
+
+    virtual void cleanup() {
+        test_ucp_wakeup::cleanup();
+        close(m_epfd);
+    }
+
+    int m_epfd;
+};
+
+UCS_TEST_P(test_ucp_wakeup_external_epollfd, epoll_wait)
+{
+    const ucp_datatype_t DATATYPE = ucp_dt_make_contig(1);
+    const uint64_t TAG = 0xdeadbeef;
+    void *req;
+
+    sender().connect(&receiver(), get_ep_params());
+
+    uint64_t send_data = 0x12121212;
+    req = ucp_tag_send_nb(sender().ep(), &send_data, sizeof(send_data), DATATYPE,
+                          TAG, send_completion);
+    if (UCS_PTR_IS_PTR(req)) {
+        wait(req);
+    } else {
+        ASSERT_UCS_OK(UCS_PTR_STATUS(req));
+    }
+
+    uint64_t recv_data = 0;
+    req = ucp_tag_recv_nb(receiver().worker(), &recv_data, sizeof(recv_data),
+                          DATATYPE, TAG, (ucp_tag_t)-1, recv_completion);
+    while (!ucp_request_is_completed(req)) {
+
+        ucp_worker_h recv_worker = receiver().worker();
+
+        if (ucp_worker_progress(recv_worker)) {
```

Comment:
would not this complete the test in most of the cases?

Reply:
no, in most cases (at least in IB), it takes some time for the packet to be looped-back and received. this time is longer than what it takes to go from "ucp_tag_send_nb" to the code line above.

Reply:
ok